dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,301 @@
+    std::wregex regex(L"^-?([^-+/*\\(\\)\\^\\s]+)");
+    std::wsmatch match;
+    if (std::regex_search(input.cbegin(), input.cend(), match, regex))
+    {
+        // Might be a number
+        const auto matchLength = match[0].length();
+        const auto parsedNum = numberParser.ParseDouble(input.substr(0, matchLength));
+
+        if (parsedNum)
+        {
+            // Parsing was successful
+            outValue = parsedNum.Value();
+            charLength = matchLength;
+        }
+    }
+
+    return charLength;
+}
+
+int NumberBoxParser::GetPrecedenceValue(wchar_t c)
+{
+    int opPrecedence = 0;
+    if (c == L'*' || c == L'/')
+    {
+        opPrecedence = 1;
+    }
+    else if (c == L'^')
+    {
+        opPrecedence = 2;
+    }
+
+    return opPrecedence;
+}
+
+// Converts a list of tokens from infix format (e.g. "3 + 5") to postfix (e.g. "3 5 +")
+std::vector<MathToken> NumberBoxParser::ConvertInfixToPostfix(std::vector<MathToken> infixTokens)
+{
+    std::vector<MathToken> postfixTokens;
+    std::stack<MathToken> operatorStack;
+
+    bool error = false;
+
+    for (auto const token : infixTokens)
+    {
+        if (token.Type == MathTokenType::Numeric)
+        {
+            postfixTokens.push_back(token);
+        }
+        else if (token.Type == MathTokenType::Operator)
+        {
+            while (!operatorStack.empty())
+            {
+                const auto top = operatorStack.top();
+                if (top.Type != MathTokenType::Parenthesis && (GetPrecedenceValue(top.Char) >= GetPrecedenceValue(token.Char)))
+                {
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+                else
+                {
+                    break;
+                }
+            }
+            operatorStack.push(token);
+        }
+        else if (token.Type == MathTokenType::Parenthesis)
+        {
+            if (token.Char == L'(')
+            {
+                operatorStack.push(token);
+            }
+            else
+            {
+                while (!operatorStack.empty() && operatorStack.top().Char != L'(')
+                {
+                    // Pop operators onto output until we reach a left paren
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+
+                if (operatorStack.empty())
+                {
+                    // Broken parenthesis
+                    error = true;
+                    break;
+                }
+
+                // Pop left paren and discard
+                operatorStack.pop();
+            }
+        }
+    }
+
+    // Pop all remaining operators.
+    while (!operatorStack.empty())
+    {
+        if (operatorStack.top().Type == MathTokenType::Parenthesis)
+        {
+            // Broken parenthesis
+            error = true;
```

Comment:
>error = true; [](start = 12, length = 13)

There is no way to correct for an error so instead of tracking state and continuing could we just early return the error empty vector?

Reply:
I thought early returns were forbidden?

Reply:
hmm, i think it is discouraged, but not returning here seems like it could be a waste, you might parse the remainder of a huge expression that you knew early on would be an error.

---
In reply to: [345448495](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345448495) [](ancestors = 345448495)

Reply:
I totally agree so if you think it's cool I'm going to do that.